### PR TITLE
server: adds truncated exponential backoff + rand for retry

### PR DESCRIPTION
**Pull request by cfromknecht**
This commit adds a backoff policy to the peer termination watcher to avoid getting stuck in tight connection loops with failing peers. The maximum backoff is currently set to 128s, and each backoff is randomized so that two instances using the same algorithm have some hope of desynchronizing. Note, a peer's backoff will only be reset if we manually disconnect via `DisconnectPeer` or restart the server.

### CODE REVIEW

File: lntest/harness.go [hunk 1]
```diff
@@ -281,6 +281,7 @@ func (n *NetworkHarness) ConnectNodes(ctx context.Context, a, b *HarnessNode) er
 			Pubkey: bobInfo.IdentityPubkey,
 			Host:   b.cfg.P2PAddr(),
 		},
+		Perm: true,
```
Comment (Roasbeef) [inline lntest/harness.go]: Not sure if this should be set to perm, as there're conditions where we want to connect peers in the tests, but then after a restart they shouldn't attempt to automatically reconnect.

File: server.go [hunk 2]
```diff
@@ -42,6 +44,14 @@ var (
 	// ErrServerShuttingDown indicates that the server is in the process of
 	// gracefully exiting.
 	ErrServerShuttingDown = errors.New("server is shutting down")
+
+	// defaultBackoff is the starting point for exponential backoff for
+	// reconnecting to persistent peers.
+	defaultBackoff = time.Millisecond
```
Comment (Roasbeef) [inline server.go]: Seems too low for an initial starting value?

File: server.go [hunk 3]
```diff
@@ -42,6 +44,14 @@ var (
...
+
+	// maximumBackoff is the largest backoff we will permit when
+	// reattempting connections to persistent peers.
+	maximumBackoff = 128 * time.Second
```
Comment (Roasbeef) [inline server.go]: Why not just use `time.Minute` here instead?

File: server.go (hunk 2 above)
Comment (aakselrod) [reply server.go]: Agreed, 1s might be a good starting point.

File: lntest/harness.go (hunk 1 above)
Comment (aakselrod) [reply lntest/harness.go]: Agreed here.

File: server.go [hunk 4]
```diff
@@ -1232,24 +1253,39 @@ func (s *server) peerTerminationWatcher(p *peer) {
 			return
 		}
 
-		srvrLog.Debugf("Attempting to re-establish persistent "+
-			"connection to peer %v", p)
-
-		// If so, then we'll attempt to re-establish a persistent
-		// connection to the peer.
+		// Otherwise, we'll launch a new connection requests in order to
```
Comment (aakselrod) [inline server.go]: Nit: "request" should be singular.

File: lntest/harness.go (hunk 1 above)
Comment (cfromknecht) [reply lntest/harness.go]: I mainly added this to be sure this doesn't get us stuck in any weird states, the itests seem to handle it just fine. Will revert.

File: server.go (hunk 2 above)
Comment (cfromknecht) [reply server.go]: Originally had it at 1 second but was failing itests even without perm. Swapping back now seems to be fine tho lol

File: server.go (hunk 3 above)
Comment (cfromknecht) [reply server.go]: I like base 2 and I cannot lie

**Comment by cfromknecht** [comment]
Thank you @aakselrod and @Roasbeef for the reviews, I've updated the default backoff to 1s, the maximum backoff to 1m, corrected the grammatical error, and removed the use of Perm from the lntest package! ⚡️

**Comment by Roasbeef** [comment]
Can you rebase this? Thanks!

**Pull request merged**